File: Python/src/quick_sort.py
from typing import List


def quick_sort(arr: List[int]) -> List[int]:
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]

    left = [n for n in arr if n < pivot]
    middle = [n for n in arr if n == pivot]
    right = [n for n in arr if n > pivot]

    return quick_sort(left) + middle + quick_sort(right)


if __name__ == '__main__':
    arr = [3, 6, 8, 10, 1, 2, 1]
    print(f"Unsorted Array: {arr}")

    print(f"Sorted Array: {quick_sort(arr)}")

File: JavaScript/quick_sort.js
const swap = (arr, left, right) =>
    ([arr[left], arr[right]] = [arr[right], arr[left]]);

function partition(arr, left, right) {
    const pivotValue = arr[right];

    let i = left - 1;

    for (let j = left; j < right; j++) {
        if (arr[j] <= pivotValue) {
            i += 1;
            swap(arr, i, j);
        }
    }

    swap(arr, i + 1, right);
    return i + 1;
}

function quickSort(arr, left, right) {
    if (left < right) {
        let pivot = partition(arr, left, right);

        quickSort(arr, left, pivot - 1);
        quickSort(arr, pivot + 1, right);
    }
}

const arr = [1, 7, 4, 1, 10, 9, -2];
console.log(`Unsorted Array: ${arr}`);

quickSort(arr, 0, arr.length - 1);
console.log(`Sorted Array: ${arr}`);
